src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -494,6 +503,166 @@ private static SetTargetResult TrySetTargetField(ref GenerationContext context,
             return SetTargetResult.Missing;
         }
 
+        private static bool IsValidInitialTypeForReplacementType(TypeDesc initialType, TypeDesc replacementType)
+        {
+            if (replacementType.IsByRef)
+            {
+                if (!initialType.IsByRef)
+                {
+                    return false;
+                }
+
+                return IsValidInitialTypeForReplacementType(((ByRefType)initialType).ParameterType, ((ByRefType)replacementType).ParameterType);
+            }
+
+            if (replacementType.IsPointer)
+            {
+                return initialType is PointerType { ParameterType.IsVoid: true };
+            }
+
+            Debug.Assert(!replacementType.IsValueType);
+
+            return initialType.IsObject;
+        }
+
+        private static SetTargetResult TrySetTargetMethodSignature(ref GenerationContext context)
+        {
+            EcmaMethod method = context.Declaration;
+            MetadataReader reader = method.MetadataReader;
+            MethodDefinition methodDef = reader.GetMethodDefinition(method.Handle);
+            ParameterHandleCollection parameters = methodDef.GetParameters();
+
+            MethodSignature originalSignature = method.Signature;
+
+            MethodSignatureBuilder updatedSignature = new MethodSignatureBuilder(originalSignature);
+
+            foreach (ParameterHandle parameterHandle in parameters)
+            {
+                Parameter parameter = reader.GetParameter(parameterHandle);
+                CustomAttribute? unsafeAccessorTypeAttribute = null;
+                foreach (CustomAttributeHandle customAttributeHandle in parameter.GetCustomAttributes())
+                {
+                    reader.GetAttributeNamespaceAndName(customAttributeHandle, out StringHandle namespaceName, out StringHandle name);
+                    if (reader.StringComparer.Equals(namespaceName, "System.Runtime.CompilerServices")
+                        && reader.StringComparer.Equals(name, "UnsafeAccessorTypeAttribute"))
+                    {
+                        unsafeAccessorTypeAttribute = reader.GetCustomAttribute(customAttributeHandle);
+                        break;
+                    }
+                }
+
+                if (unsafeAccessorTypeAttribute is null)
+                {
+                    continue;
+                }
+
+                if (parameter.SequenceNumber > originalSignature.Length)
+                {
+                    return SetTargetResult.Invalid;
+                }
+
+                bool isReturnValue = parameter.SequenceNumber == 0;
+
+                if (isReturnValue && context.Kind is UnsafeAccessorKind.Field or UnsafeAccessorKind.StaticField)
+                {
+                    // We can't support UnsafeAccessorTypeAttribute on fields
+                    // today as it would create a type-safety hole.
```

Comment:
This problem is not limited to fields. It applies to methods that return byrefs too.

E.g. this should be blocked for the same reason fields are blocked:
```
using System.Runtime.CompilerServices;

unsafe class Test
{
    static void Main()
    {
        ref object ro = ref MProxy(null);

        ro = "Hello";

        Console.WriteLine(g.GetType());
    }

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name="M")]
    [return: UnsafeAccessorType("Test&")] 
    extern static ref object MProxy(Test t);

    static ref Test M() => ref g;

    static Test g;
}
```

Reply:
Update NativeAOT to block ref-returning methods as well. Leaving this unresolved as CoreCLR, Mono, and tests need to be updated.

Reply:
Test added.